### PR TITLE
feat(port): Port the Dash spell effect from DDA

### LINKED ISSUE #4534: Port Request: "Dash" Spell Effect
### Is your feature request related to a problem? Please describe.

We, as of yet, do not have any target teleport / movement effects in the spellcode other than the janky work-around of SWAP_POS flag. This is a fairly big limitation on the spells we can implement.

### Describe the solution you'd like

Port the spell effect "Dash" from DDA, originally found in [this DDA PR](https://github.com/CleverRaven/Cataclysm-DDA/pull/44357).

### Additional context

I'm somewhat surprised this wasn't already ported, especially given we've tried to port bits of it before.

### PR DESCRIPTION
**Pull request by RobbieNeko**
### Required

- [x] I wrote the PR title in [conventional commit format](https://docs.cataclysmbn.org/en/contribute/changelog_guidelines/).
- [x] I ran the [code formatter](https://docs.cataclysmbn.org/en/contribute/contributing/#code-style).
- [x] I linked any relevant issues using [github keyword syntax](https://docs.cataclysmbn.org/en/contribute/contributing/#pull-request-notes) so it can be closed automatically.
- [x] I have [committed my changes to new branch that isn't `main`](https://docs.cataclysmbn.org/en/contribute/contributing/#make-your-changes) so it won't cause conflict when updating `main` branch later.

### Optional

- [x] This PR ports commits from DDA or other cataclysm forks.
  - [x] I have attributed original authors in the commit messages adding [`Co-Authored-By`](https://docs.github.com/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) in the commit message.
  - [x] I have linked the URL of original PR(s) in the description.
- [x] This is a C++ PR that modifies JSON loading or behavior.
  - [x] I have documented the changes in the appropriate location in the `doc/` folder.
  - [x] If documentation for this feature does not exist, please write it or at least note its lack in PR description.

## Purpose of change

Fixes #4534 
Controlled teleportation without the jank of swap_pos has been a major wishlist item by several modders dealing with spell code. This solves this big wishlist item by porting the dash spell effect from DDA, which is generally exactly what we were looking for. Now, mods can be free of the jank that comes with the swap_pos spell flag.

## Describe the solution

Ports https://github.com/CleverRaven/Cataclysm-DDA/pull/44357 from DDA

Remove the code that tries to damage things because it never worked anyway (even in DDA) and is potentially messing up stuff regarding auto-targeting and enraging NPCs. If you want a damaging effect, just use the `"extra-effects":` field to put a cone or other damaging spell on the end of it yourself (and it'll have a better chance of working).

## Describe alternatives you've considered

Reinvent the wheel

## Testing

It builds on my end, and it works as expected for dashing and for leaving behind a trail of fields!

image

## Additional context

First (non-test related) c++ PR for me!

### CODE REVIEW

**Review by scarf005 — DISMISSED** [review]
@chaosvolt could you playtest just in case?

**Review by chaosvolt — COMMENTED** [review]
1. Compiled and load-tested.
2. Added in debug spell to Magiclysm.
3. Learned it and boosted its level to max.
4. Seems to work as expected for the most part, however...

Dashing into an enemy stops you early and deals no damage as evidently expected, however the spell auto-targets neutral NPCs and casting it at them will aggro them despite doing nothing.

Bonus, selecting a tile behind an NPC dashes into them, still harmlessly, but doesn't aggro.

**Comment by RobbieNeko** [comment]
Hmm... unfortunately, I'm not sure exactly what I can do with regards to that. My best guess would be that the attempt to do damage at the end of it *is* getting interpreted by the npcs, even if it's not actually going so far as to damage them.

**Comment by RobbieNeko** [comment]
I think I might just straight up remove the damage aspects, it seems like it's just getting in the way of what should be a movement spell effect

**Comment by RobbieNeko** [comment]
Removed the damaging parts of the function, hopefully that'll make npcs care less about dashing players (and if you want your dash to deal damage anyway, just stick the 'cast another spell' flag onto your spell and make it cast a cone or whatever other spell you desire.)

File: data/mods/Magiclysm/Spells/debug.json [hunk 1]
```diff
@@ -230,5 +230,27 @@
...
+    "id": "debug_dash",
+    "name": "Debug Dash",
+    "description": "You can teleport instantaneously forward.",
+    "valid_targets": [ "ground", "hostile" ],
+    "effect": "dash",
+    "//": "As of writing, damage doesn't work but it's good to keep it for when it does",
+    "min_damage": 80,
+    "max_damage": 80,
+    "damage_increment": 10,
+    "damage_type": "bash",
```
Comment (scarf005) [inline data/mods/Magiclysm/Spells/debug.json:245]:
```suggestion
```

might as well remove this line if you decide to remove spell damages

File: data/mods/Magiclysm/Spells/debug.json [hunk 2]
```diff
@@ -230,5 +230,27 @@
...
+    "field_id": "fd_fire",
+    "min_field_intensity": 2,
+    "max_field_intensity": 2,
```
Comment (scarf005) [inline data/mods/Magiclysm/Spells/debug.json:248]: does leaving fields work properly? in the screenshot i was unable to see fires

Comment (RobbieNeko) [reply data/mods/Magiclysm/Spells/debug.json:248]: Sorry, that screenshot was before I added the fields. yes, the fields work just fine

Comment (RobbieNeko) [reply data/mods/Magiclysm/Spells/debug.json:243]: image

**Review by scarf005 — APPROVED** [review]

**Pull request merged**